### PR TITLE
sql-parser: parse CREATE/DROP ROLE

### PR DESCRIPTION
**Pull request by benesch**
Add support for parsing a restricted subset of PostgreSQL's CREATE ROLE
and DROP ROLE statements. These statements do not yet do anything in
Materialize.

See #4626 for context. The basic idea is to support a very restricted
form of user accounts to facilitate securing the network interface.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5468)

### CODE REVIEW

File: src/sql-parser/src/parser.rs [hunk 1]
```diff
@@ -1721,6 +1724,32 @@ impl<'a> Parser<'a> {
...
+    fn parse_create_role(&mut self) -> Result<Statement<Raw>, ParserError> {
+        let is_user = match self.expect_one_of_keywords(&[ROLE, USER])? {
+            ROLE => false,
+            USER => true,
+            _ => unreachable!(),
+        };
+        let name = self.parse_identifier()?;
+        let _ = self.parse_keyword(WITH);
+        let mut options = vec![];
+        loop {
```
Comment (sploiselle) [inline src/sql-parser/src/parser.rs:1736]: idk if this matters but pg seemingly does type-like parsing for these options and only parses those with > 1 token:

```
sean=#  CREATE ROLE usr WITH badopt ;
ERROR:  unrecognized role option "badopt"

sean=#  CREATE ROLE usr WITH CONNECTION;
ERROR:  syntax error at or near ";"
LINE 1: CREATE ROLE usr WITH CONNECTION;
```

**Review by sploiselle — APPROVED** [review]
LGTM

File: src/sql-parser/src/parser.rs (hunk 1 above)
Comment (sploiselle) [reply src/sql-parser/src/parser.rs:1736]: Sorry––wrote that right before bed. Not very clear!

If the text following `WITH` is not a token known to be part of a multi-token option, PG parses it as an option––the content of the options are then validated later. The consequence of this is that `SUPERUSER` et. al. are not tokens within PG, and checking for valid options isn't done in parsing, but instead in planning.

Don't think it matters much––I just know how much you love minimizing the number of unnecessary tokens we have strewn about.

**Comment by JLDLaughlin** [comment]
Do we have plans to do something with roles in the near future?

**Comment by benesch** [comment]
We do! https://github.com/MaterializeInc/materialize/issues/4626#issuecomment-767979363

File: src/sql-parser/src/parser.rs (hunk 1 above)
Comment (benesch) [reply src/sql-parser/src/parser.rs:1736]: Ohh, I see! Good point. I'm not terribly worried about these extra keywords (nonrestricted keywords in our SQL parser are essentially zero cost, which is not the case in PostgreSQL) but definitely possible that we'll need to clean this up in the future.

**Comment by benesch** [comment]
Oops. Linked you to the wrong comment. https://github.com/MaterializeInc/materialize/issues/4626#issuecomment-767878604

**Pull request merged**